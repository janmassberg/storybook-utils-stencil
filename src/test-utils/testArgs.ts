export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: (): void => {
        // Implementation
    },
    onTestEvent: (event: Event): void => {
        // Handle event
        console.log(event.type);
    },
    _slot: "Test children",
};

export const testArgsButton = {
    required: true,
    variant: "primary",
    onClick: (event: Event) => {
        // Handle click event
        console.log(event.target);
    },
    _slot: "Label",
};

export const testArgsComponentGenerator = {
    component: "my-component",
    args: testArgs,
};

export const testCopmonentSlot = `<my-child-component foo-bar="foo">
<my-nested-child-component bar-foo="bar"><span class="test-class">Label</span></my-nested-child-component>
</my-child-component>`;
