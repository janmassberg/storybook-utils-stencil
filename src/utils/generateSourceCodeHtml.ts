import { toKebabCase } from "./stringUtils";
import { formatHtml } from "./formatHtml";
import { filterHtmlAttributes, ArgsType } from "./argUtils";

const getHtmlAttributes = (attributes: ArgsType): string[] => {
    return attributes
        .map(([key, value]) => {
            key = toKebabCase(key);
            if (typeof value === "boolean") {
                return value ? key : "";
            }
            return `${key}="${value}"`;
        })
        .filter((str) => str !== "");
};

export const generateSourceCodeHtml = (
    component: string,
    args: any
): string => {
    const htmlAttributes = getHtmlAttributes(filterHtmlAttributes(args));
    const children = args._children || "";
    return formatHtml(
        `<${component} ${htmlAttributes.join(" ")}>${children}</${component}>`
    );
};
