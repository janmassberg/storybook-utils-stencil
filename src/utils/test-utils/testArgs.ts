export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: () => {},
    onTestEvent: () => {},
    _children: "Test children",
};
