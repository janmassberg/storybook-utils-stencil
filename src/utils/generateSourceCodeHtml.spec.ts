import { generateSourceCodeHtml } from "./generateSourceCodeHtml";
import { testArgs } from "./test-utils/testArgs";

describe("generateSourceCodeHtml", () => {
    it("should build the HTML source for a component with args correctly", () => {
        expect(
            generateSourceCodeHtml("my-component", testArgs)
        ).toMatchSnapshot();
    });
});
