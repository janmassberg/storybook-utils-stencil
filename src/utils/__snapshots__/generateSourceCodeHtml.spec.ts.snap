// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSourceCodeHtml should build the HTML source for a component with args correctly 1`] = `
"<my-component test-true test-string=\\"test-string\\" test-number=\\"12345\\">
  Test children
</my-component>
"
`;
